cocotb/bfms.py

```diff
@@ -0,0 +1,206 @@
+        
+        
+        for i in range(len(args)):
+            a = args[i]
+            t = signature[i]
+            self.signature.append(BfmMethodParamInfo(a, t))
+            try:
+                import simulator
+                if isinstance(t, bfm_param_int_t):
+                    if t.s:
+                        self.type_info.append(simulator.BFM_SI_PARAM)
+                    else:
+                        self.type_info.append(simulator.BFM_UI_PARAM)
+            except Exception:
+                # When we're not running in simulation, don't 
+                # worry about being able to access constants from simulation
+                self.type_info.append(None)
+
+
+class BfmTypeInfo():
+    
+    def __init__(self, T, hdl, import_info, export_info):
+        self.T = T
+        self.hdl = hdl
+        self.import_info = import_info
+        self.export_info = export_info
+
+class BfmInfo():
+    
+    def __init__(self, bfm, id, inst_name, type_info):
+        self.bfm = bfm
+        self.id = id
+        self.inst_name = inst_name
+        self.type_info = type_info
+        
+    def call_method(self, method_id, params):
+        self.type_info.export_info[method_id].T(
+            self.bfm, *params)
+
+class BfmMgr():
+    
+    m_inst = None
+    
+    def __init__(self):
+        self.bfm_l = []
+        self.bfm_type_info_m = {}
+        self.m_initialized = False
+    
+    def add_type_info(self, T, type_info):
+        self.bfm_type_info_m[T] = type_info
+    
+    @staticmethod
+    def get_bfms():
+        return BfmMgr.inst().bfm_l
+
+    @staticmethod    
+    def find_bfm(path_pattern):
+        inst = BfmMgr.inst()
+        bfm = None
+       
+        path_pattern_re = re.compile(path_pattern)
+        
+        for b in inst.bfm_l:
+            if path_pattern_re.match(b.bfm_info.inst_name):
+                bfm = b
+                break
+        
+        return bfm
+   
+    @staticmethod
+    def inst():
+        if BfmMgr.m_inst is None:
+            BfmMgr.m_inst = BfmMgr()
+            
+        return BfmMgr.m_inst
+   
+    def load_bfms(self):
+        '''
+        Obtain the list of BFMs from the native layer
+        '''
+        import simulator
+        n_bfms = simulator.bfm_get_count()
+        for i in range(n_bfms):
+            info = simulator.bfm_get_info(i)
+            instname = info[0]
+            clsname = info[1]
+            if clsname.find('.') == -1:
+                raise Exception("Incorrectly-formatted BFM class name {!r}".format(clsname))
+            pkgname = clsname[:clsname.rfind('.')]
+            clsleaf = clsname[clsname.rfind('.')+1:]
+          
+            try: 
+                pkg = importlib.import_module(pkgname)
+            except Exception as e:
+                raise Exception("Failed to import BFM package \"" + pkgname + "\"")
```

Comment:
```suggestion
                raise Exception("Failed to import BFM package {!r}".format(pkgname))
```
will add the quotes for you. Same throughout